### PR TITLE
DI-176 Dev pipeline definition

### PR DESCRIPTION
**Pull request by ed-sparkes**
## Link to JIRA Ticket

- Create Development Pipeline for the Master Branch

## Description

Terraform scripts and updates to MakeFiles to provision CodePipeline for Continuous integration and deployment to Dev environment. 

Build has the Following steps

- Trigger from change to Master 
- Run unit tests and publish test report
- Build components
- Run Component tests and publish test report
- Publish components
- Deploy

### Noteworthy Changes

- add `docker-hub-signin` make target to avoid rate limiting errors from docker hub for repeated pullls of nhsd/tools
- add junit output to both unit and component tests and integrate with code build so test results visible in AWS console
- new ct.mk file for running component tests on linux build agent
- downgrade docker compose to 3.8 as 3.9 was not supported on the build agent 
- remove the dependson from docker compose as was not supported by build agent 
- replace `ARG FUNCTION_DIR="/var/task"` with `/var/task` repeated where needed inline as the variable scoping was not working correctly on the version of docker on the build agent
-  add `RUN apk add python3 make gcc g++` to serverless DockerFile to enable rebuild of node-gyp 
- 
## Type of change

- New feature (non-breaking change which adds functionality)

## Testing

Please tick the testing that has been completed

- [ ] Unit tests
- [ ] Component tests

## Developer Checklist

- [x] I have performed a self-review of my own code
- [ ] I have run the [code formatting checks](../README.md#code-quality)
- [ ] I have run the [code quality checks](../README.md#code-quality)
- [ ] New code meets [standards](https://nhsd-confluence.digital.nhs.uk/display/DI/DI+Ways+of+Working) agreed by the team
- [ ] Unit test code coverage is at or above 80%
- [ ] New and existing unit tests pass locally with my changes
- [ ] Tests have added that prove my fix is effective or that my feature works (Component tests)
- [ ] I have made corresponding changes to the documentation
- [ ] I have cleaned down my environment (if created)

## Code Reviewer Checklist

- [ ] I have run the unit tests and they run correctly
- [ ] I can confirm the changes have been tested or approved by a tester
- [ ] I can confirm no remaining infrastructure is left over from this branch

### CODE REVIEW

File: infrastructure/stacks/development-pipeline/codepipeline.tf
Comment (JackPlowman) [inline infrastructure/stacks/development-pipeline/codepipeline.tf:18]: Is there a particular version for this being version 1 and not version 2?

File: infrastructure/stacks/development-pipeline/component-test-buildspec.yml
Comment (JackPlowman) [inline infrastructure/stacks/development-pipeline/component-test-buildspec.yml:1]: What's the point of the component test build spec if it isn't used?

File: infrastructure/stacks/development-pipeline/locals.tf
Comment (JackPlowman) [inline infrastructure/stacks/development-pipeline/locals.tf:1]: This file can be removed and the providers.tf can be updated to the code below. This means that you don't have to tag every resource
```
provider "aws" {
  profile = var.aws_profile
  region  = var.aws_region

  default_tags {
    tags = {
      "Programme"   = "uec"
      "Service"     = "uec-dos-int"
      "Product"     = "uec-dos-int"
      "Environment" = var.profile
    }
  }
}
```

File: infrastructure/stacks/development-pipeline/unit-tests-buildspec.yml
Comment (JackPlowman) [inline infrastructure/stacks/development-pipeline/unit-tests-buildspec.yml:56]: Can the results reports which are commented out be removed

File: infrastructure/stacks/development-pipeline/variables.tf
Comment (JackPlowman) [inline infrastructure/stacks/development-pipeline/variables.tf:60]: Can the uppercase  `AWS_ACCOUNT` variables be removed and instead use the lowercase versions `aws_account`. As the lowercase variables are automatically exported so don't have to be manually exported.

**Review by JackPlowman — CHANGES_REQUESTED** [review]

File: build/automation/var/profile/dev.mk
Comment (ed-sparkes) [reply build/automation/var/profile/dev.mk:61]: yup will put back once completed testing the new db integration

File: infrastructure/stacks/development-pipeline/codebuild.tf
Comment (ed-sparkes) [reply infrastructure/stacks/development-pipeline/codebuild.tf:32]: have put something in here but suggest we review in a separate optimisation ticket so we can get this merged and working

File: infrastructure/stacks/development-pipeline/codepipeline.tf
Comment (ed-sparkes) [reply infrastructure/stacks/development-pipeline/codepipeline.tf:18]: Moved to version 2 which required us to change provider to CodeStarSourceConnection version 1 confusingly :)

**Review by JackPlowman — APPROVED** [review]
LGMT 👍

File: infrastructure/stacks/development-pipeline/unit-tests-buildspec.yml
Comment (mabe13) [inline infrastructure/stacks/development-pipeline/unit-tests-buildspec.yml:56]: Once we're finished with developing this pipeline we should remove the commented out sections from build spec yml files

File: infrastructure/stacks/development-pipeline/build-buildspec.yml
Comment (JackPlowman) [inline infrastructure/stacks/development-pipeline/build-buildspec.yml:24]: Do you need to install docker here if you are also doing it below?

**Review by JackPlowman — CHANGES_REQUESTED** [review]
One small thing I noticed after a second look over. Otherwise all good.

File: infrastructure/stacks/development-pipeline/unit-tests-buildspec.yml
Comment (mabe13) [reply infrastructure/stacks/development-pipeline/unit-tests-buildspec.yml:56]: I think if we're going to make those kinda of pipeline changes fairly soon, then it's ok to leave the comment out parts, but if it's more long term then it would be better to remove it. Removing the comment would make it more readable, and we can always look up the build-spec stuff in the future

That said don't mind, I'll leave it to you to decide if it's need

**Comment by JackPlowman** [comment]
@ed-sparkes Once this is rebase I am happy to approve this PR

**Review by JackPlowman — APPROVED** [review]
LGTM 👍

**Review by mabe13 — APPROVED** [review]
LGTM 👍

**Pull request merged**
